PyRDF/Proxy.py

```diff
@@ -1,54 +1,90 @@
 from __future__ import print_function
 from PyRDF.CallableGenerator import CallableGenerator
+from PyRDF.Operation import Operation
+from PyRDF.Node import Node
 
 
 class Proxy(object):
     """
-    Instances of Proxy act as futures of the result produced
-    by some action. They implement a lazy synchronization
-    mechanism, i.e., when they are accessed for the first time,
-    they trigger the execution of the whole RDataFrame graph.
-
-    Attributes
-    ----------
-    backend
-        A class member to store a backend object
-        based on the configuration set by the user.
-
-    action_node
-        The action node that the current Proxy
-        instance wraps.
-
+    Abstract class for proxies objects. These objects help to keep track of
+    nodes' variable assignment. That is, when a node is no longer assigned
+    to a variable by the user, the role of the proxy is to show that. This is
+    done via changing the value of the `has_user_references` of the proxied
+    node from `True` to `False`.
     """
-
-    def __init__(self, action_node):
+    def __init__(self, node):
         """
-        Creates a new `Proxy` object for a
-        given action node.
+        Creates a new `Proxy` object for a given node.
 
         Parameters
         ----------
-        action_node : PyRDF.Node
-            The action node that the current Proxy
-            should wrap.
-
+        proxied_node : PyRDF.Node
+            The node that the current Proxy should wrap.
         """
-        self.action_node = action_node
+        self.proxied_node = node
 
     def __getattr__(self, attr):
         """
-        Intercepts calls on the result of
-        the action node.
-
-        Returns
-        -------
-        function
-            A method to handle an operation call to the
-            current action node.
+        Intercepts calls to attributes and methods of the proxied node and
+        returns the appropriate object(s).
 
+        Parameters
+        ----------
+        attr : str
+            The name of the attribute or method of the proxied node the user
+            wants to access.
+        """
+        # Old: some tests call dunder methods on nodes so
+        # proxies have to return dunder methods too.
+        # Check if the current call is a dunder method call
+        # import re
+        # if re.search("^__[a-z]+__$", attr):
+        #     # Raise an AttributeError for all dunder method calls
+        #     raise AttributeError("Such an attribute is not set ! ")
+
+        # if the current node is an action and the user issues a call
+        # return the result of the action
+        try:
+            if self.proxied_node.operation.is_action():
+                self._cur_op_call = attr  # Stores the name of operation call
```

Comment:
this is outdated, it refers to the commit where I removed the proxy subclasses. Now is reverted back to `_cur_attr` attribute for the `ActionProxy` class. Still don't like it though, would prefer to call it `cur_action_attr` or `cur_action_call` because it's the string we pass to the `GetValue` method.